File: claudesync/gitutil.py
"""Small wrapper around the git CLI used to version backups."""

from __future__ import annotations

import subprocess
from pathlib import Path


class GitError(RuntimeError):
    pass


def _run(args: list[str], cwd: Path) -> subprocess.CompletedProcess:
    return subprocess.run(
        ["git", *args],
        cwd=cwd,
        capture_output=True,
        text=True,
    )


def is_repo(path: Path) -> bool:
    return (path / ".git").is_dir()


def init(path: Path) -> None:
    result = _run(["init"], path)
    if result.returncode != 0:
        raise GitError(result.stderr.strip())


def add_all(path: Path) -> None:
    result = _run(["add", "-A"], path)
    if result.returncode != 0:
        raise GitError(result.stderr.strip())


def has_staged_changes(path: Path) -> bool:
    result = _run(["diff", "--cached", "--quiet"], path)
    # exit code 1 means there ARE staged changes, 0 means none
    return result.returncode == 1


def commit(path: Path, message: str) -> bool:
    """Commit staged changes. Returns False if there was nothing to commit."""
    if not has_staged_changes(path):
        return False
    result = _run(["commit", "-m", message], path)
    if result.returncode != 0:
        raise GitError(result.stderr.strip())
    return True


def set_remote(path: Path, remote_url: str, name: str = "origin") -> None:
    existing = _run(["remote"], path).stdout.split()
    if name in existing:
        _run(["remote", "set-url", name, remote_url], path)
    else:
        _run(["remote", "add", name, remote_url], path)


def push(path: Path, remote: str = "origin", branch: str | None = None) -> None:
    args = ["push", "-u", remote]
    if branch:
        args.append(branch)
    result = _run(args, path)
    if result.returncode != 0:
        raise GitError(result.stderr.strip())


def current_branch(path: Path) -> str:
    result = _run(["rev-parse", "--abbrev-ref", "HEAD"], path)
    branch = result.stdout.strip()
    return branch if branch and branch != "HEAD" else "main"

File: desktop/src/core/gitutil.js
'use strict';

const { spawnSync } = require('child_process');

class GitError extends Error {}

function run(args, cwd) {
  return spawnSync('git', args, { cwd, encoding: 'utf8' });
}

function isRepo(dir) {
  const fs = require('fs');
  const path = require('path');
  try {
    return fs.statSync(path.join(dir, '.git')).isDirectory();
  } catch {
    return false;
  }
}

function init(dir) {
  const result = run(['init'], dir);
  if (result.status !== 0) throw new GitError(result.stderr.trim());
}

function addAll(dir) {
  const result = run(['add', '-A'], dir);
  if (result.status !== 0) throw new GitError(result.stderr.trim());
}

function hasStagedChanges(dir) {
  const result = run(['diff', '--cached', '--quiet'], dir);
  return result.status === 1;
}

/** Commits staged changes. Returns false if there was nothing to commit. */
function commit(dir, message) {
  if (!hasStagedChanges(dir)) return false;
  const result = run(['commit', '-m', message], dir);
  if (result.status !== 0) throw new GitError(result.stderr.trim());
  return true;
}

function setRemote(dir, remoteUrl, name = 'origin') {
  const existing = run(['remote'], dir).stdout.split(/\s+/).filter(Boolean);
  if (existing.includes(name)) {
    run(['remote', 'set-url', name, remoteUrl], dir);
  } else {
    run(['remote', 'add', name, remoteUrl], dir);
  }
}

function push(dir, { remote = 'origin', branch } = {}) {
  const args = ['push', '-u', remote];
  if (branch) args.push(branch);
  const result = run(args, dir);
  if (result.status !== 0) throw new GitError(result.stderr.trim());
}

function currentBranch(dir) {
  const result = run(['rev-parse', '--abbrev-ref', 'HEAD'], dir);
  const branch = result.stdout.trim();
  return branch && branch !== 'HEAD' ? branch : 'main';
}

module.exports = {
  GitError,
  isRepo,
  init,
  addAll,
  hasStagedChanges,
  commit,
  setRemote,
  push,
  currentBranch,
};
